### ISSUE TITLE
Images always filling space after 2.2.3 - IMPORTANT!!! CachedImage.FixedOnMeasureBehavior property

### ISSUE THREAD

**Issue by kentcb** [luberda-molinet/FFImageLoading#545]
**//EDIT by Daniel Luberda:**

It's fixed now, but you must set this:

`CachedImage.FixedOnMeasureBehavior = true;`

**I had to do this to avoid breaking behavior changes as the regression exists now for quite a few versions. We'll reenable it with next major version to avoid confusion.** 

//ORIGINAL: 

Hi,

Before I go to the effort of providing a lot of detail on this, I wanted to put feelers out first. I've been hanging back on 2.2.3 for some time now because when I upgrade, my images in XF appear way too big. I assumed this was a bug in FFIL when I first upgraded (I believe it was 2.2.4), but since it's been that way for a _long_ time now, I figure maybe this is intentional?

Was there some kind of default behavior change at this point? I _think_ this only affects Android btw, but would have to confirm that as part of digging deeper if/when required.

**Comment by VincentH-Net** [comment]
I also have this issue in 2.2.8 in both Android and iOS. Replacing all standard Xamarin Forms Image controls with CachedImage results in way too big images.
Specifics:
- I'm not specifying the image size in the layout; the standard Image control uses the PNG's size
- The app has each image in multiple resolutions in the app resources in both Android and iOS  

**-> Is it intentional or a bug that CachedImage behaves differently than Image given identical properties and layouts?
-> Is there a set of CachedImage property settings to make it compatible with standard Image?**

Basically I just want to switch to CachedImage for added performance, stability and functionality. I don't want to have to change and retest all my existing Forms layouts.

An option for compatible behavour would promote adoption of FFImageLoading; I imagine many apps start out with standard image and when the app grows in size / usage want to switch with minimal rework.

**Comment by kentcb** [comment]
@VincentH-Net agreed, and your findings verify my own (I don't recall whether iOS exhibits the same problem for us, but am guessing it does if it has for you).

The thing is, 2.3.3 does have compatible behavior with a standard `Image`, so something changed thereafter, intentionally or not.

**Comment by VincentH-Net** [comment]
Agreed, Either way it has a big impact for XF devs, It is preventing me to adopt FFImageLoading, 
I hope addressing this can get some priority.

If intentional, it is a big breaking change with major impact for all FFImageLoading XF users which merits documenting a rationale including an approach how to achieve compatibility (setting?).

I hope it is just a bug,

**Comment by VincentH-Net** [comment]
I tested:
Xamarin.FFImageLoading.Forms 2.2.3 +
Xamarin.FFImageLoading 2.2.8
This shows images in the correct size on iOS and Android

-> the change/defect is in the Forms package only

**Comment by daniel-luberda** [comment]
Yes, there was a change to better support expanding layout. Here's the code responsible: 

- FFImageLoading: https://github.com/luberda-molinet/FFImageLoading/blob/master/source/FFImageLoading.Forms/CachedImage.cs#L490-L523
- Xamarin.Forms: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/Image.cs#L63-L118

Do you have any ideas to modify it to have both working as expected? You can inherit from `CachedImage` and just paste original Xamarin.Forms code, to have previous behaviour.

**Comment by kentcb** [comment]
Finally got some time to look at this today. It seems like the majority of the problems I observed when opening this issue are addressed now (using `2.2.10-pre-413`). However, there remains one issue I can spot with images that have `VerticalOptions="Center"` applied (and presumably same problem horizontally). Such images are still stretched instead of collapsing into the center of the space they occupy.

I've tried to grab the source and run my project up against it, but FFIL is refusing to build in VS2017 for me. The build batch file works, but I need to be able to debug through this to fix, and I'm not having any luck getting that far :(

**Comment by kentcb** [comment]
This is the build error I get, btw:
```
9>C:\Users\Kent\Repository\FFImageLoading\source\FFImageLoading.BaitAndSwitch\project.json(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
```
Can't see anything obviously wrong with the file in question (and I haven't changed anything since cloning the repo).

**Comment by daniel-luberda** [comment]
@kentcb Sources are now updated for VS2017. Are you able to build it now?

**Comment by kentcb** [comment]
Hmm, it's still using project.json, which is leading to this when I try and use FFIL source directly from my Xamarin project:
```
C:\Users\Kent\Repository\FFImageLoading\source\FFImageLoading.Common\project.json(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
```
Ignoring that, the good news is that I can now build FFIL in isolation from my Xamarin project. So I figured i could just reference the compiled binaries from my solution instead of using FFIL source. However, when referencing those binaries from my Xamarin project, my XAML compilation started failing (with null ref exceptions). It's not clear why, so I just disabled XAMLC to continue on.

Basically, it's really quite difficult working with FFIL as source, and I suspect the friction prevents quite a few community contributions. You should take a look at what the ReactiveUI folks (mainly @ollevett) have done to multitarget many platforms from one project. You'd wind up with 3 projects instead of `3 * $PLATFORM_COUNT`. Anyway, just a thought - you may not have the time to invest right now - completely understand if that's the case.

That said, I finally got building (against the binaries, so not a great debugging story). However, the problem is still occurring, and debugging is not really practical without source :(

It's really simple to reproduce. You just give an image more space than it needs and set it to center. A standard `Image` does the right thing, but `CachedImage` stretches the image to fill the space. Again, this regressed some time after 2.2.3.

[Here's a repro](https://github.com/luberda-molinet/FFImageLoading/files/1273897/FFILRepro.zip).

**Comment by kentcb** [comment]
UPDATE: straight after posting the above, I realised I forgot to try consuming FFIL source from my repro (as opposed to real project). Doing so works, so I'm digging deeper now.

I suspect the build errors I'm seeing are due to FFIL using legacy stuff like PCLs and project.json, whereas my real project uses netstandard and SDK-style csproj (my repro does not).

**Comment by kentcb** [comment]
OK, PR submitted (#714). Looking forward to getting this fixed @daniel-luberda. Thanks again for a great library!

**Comment by daniel-luberda** [comment]
Great! Thank you for your collaboration. I merged it, however currently that fix is disabled by default. To enable it. Set the static property: 

`CachedImage.FixedOnMeasureBehavior = true;`

I had to do this to avoid breaking behavior changes as the regression exists now for quite a few versions. We'll reenable it with next major version to avoid confusion. 


@kentcb 
- FFImageLoading projects are now .NET standard projects, earlier I was using VS2015 for building, now I switched to VS2017. 
- I enabled debug symbols for `FFImageLoading.Forms`, nice catch
- The project is already quite time consuming, so it's not easy to manage everything, but I'll try to improve project build on various environments in future. After recent changes, most of the issues should be solved anyway.

**Comment by YZahringer** [comment]
I have this exception when I set FixedOnMeasureBehavior property on UWP app in v2.2.14:
`System.TypeInitializationException: 'The type initializer for 'FFImageLoading.Forms.CachedImage' threw an exception.' 
You are referencing the Portable version in your App - you need to reference the platform (iOS/Android) version`

**Comment by daniel-luberda** [comment]
@YZahringer  Did you remove all `bin`/`obj` folders before, then clean & rebuild?

**Comment by YZahringer** [comment]
@daniel-luberda Yep, problem persist. I try reinstall of FFImageLoading nuget packages, same error. I have a NETStandard 1.4 project referenced by UWP, Droid, iOS & Mac projects. 

I use Xamarin.Forms 2.4.0.275-pre3, no problem with v2.2.13.

**Comment by daniel-luberda** [comment]
@YZahringer You must install packages to the platform specific projects. Otherwise, it won't work. Did you install the packages in a UWP specific project too?

**Comment by YZahringer** [comment]
@daniel-luberda I installed packages Xamarin.FFImageLoading.Forms, Xamarin.FFImageLoading.Svg.Forms and Xamarin.FFImageLoading.Transformations in all projects.

**Comment by daniel-luberda** [comment]
I just modified nuspec files. Could you check 2.2.15 prerelease?

**Comment by YZahringer** [comment]
@daniel-luberda It works! Thank you 👍

**Comment by EmilAlipiev** [comment]
thats weird, in my case using CachedImage.FixedOnMeasureBehavior = true; doesnt work correctly. my Images displayed super small. without using this line display it fine,

**Comment by daniel-luberda** [comment]
@EmilAlipiev 

- What platform?
- What is the original file width/ height?

**Comment by EmilAlipiev** [comment]
It is on Android. I tested using Android 6 version on a device.  here is the actual image below and my implementation in xaml. in previous versions displayed just fine with the code below and it is a bit larger but still ok using without using CachedImage.FixedOnMeasureBehavior = true

```
  <Grid   AbsoluteLayout.LayoutFlags="All"  AbsoluteLayout.LayoutBounds="0,0,1,1">
                <Grid.RowDefinitions>
                    <RowDefinition Height="Auto"></RowDefinition>
                    <RowDefinition Height="Auto"></RowDefinition>
                    <RowDefinition Height="*"></RowDefinition>              
                </Grid.RowDefinitions>
                <Grid.ColumnDefinitions>
                    <ColumnDefinition Width="*"></ColumnDefinition>
                    <ColumnDefinition Width="*"></ColumnDefinition>
                </Grid.ColumnDefinitions>


  <ffimageloading:CachedImage  x:Name="Image1"  	Aspect="AspectFit"   Grid.Row="2" Grid.Column="0"  Grid.ColumnSpan="2"  HorizontalOptions="Center" VerticalOptions="Start" 
                                        LoadingPlaceholder="loading.gif"       RetryCount="0"    CacheDuration = "120"  
                                   Source="{Binding Image1}"  
                            />
```

![image](https://user-images.githubusercontent.com/10422347/31202030-374c65b4-a961-11e7-9e58-4eba1970d4c0.png)

**Comment by daniel-luberda** [comment]
One more thing, does this behaviour differ from Xamarin.Forms `Image`?

**Comment by EmilAlipiev** [comment]
I just tested and it is also super small using XF Image.

**Comment by daniel-luberda** [comment]
As you see it's expected behaviour to be fully compatible with forms.

**Comment by EmilAlipiev** [comment]
This is weird because using AUTO and ColumnSpan=2 on datagrid I expect much larger Image. Do you think it is a bug in XF 2.4? Because I use FFImageLoading already for a long time and earlier version showed this image fine. Current version is also OK without using g `CachedImage.FixedOnMeasureBehavior = true`. I wonder what is breaking change. It looks like in 2.4 version

**Comment by daniel-luberda** [comment]
Default behaviour of XF `Image` is to show image with original file size in that scenario.

**Comment by EmilAlipiev** [comment]
Thats the problem looks like original size is much bigger than what is shown. I will write on Xamarin forms also

**Comment by chrisfoulds** [comment]
How do I specify this in XAML in Forms ? , it doesn't seem to be an available property

**Comment by chrisfoulds** [comment]
FixedOnMeasureBehavior is not a bindable property in forms so how do I use it ?

**Comment by chrisfoulds** [comment]
Ok so, I WAS using ffimageloading for my local images as my lists contain loads of images.
I have just updated by a few versions to the latest and my layouts were broke.
I added to my App.cs
CachedImage.FixedOnMeasureBehavior = true;
CachedImage.FixedAndroidMotionEventHandler = true;

They are still broke.
I replaced ffimageloading in my XAML for <Image> and they all display correctly again.
am I missing something ?

**Comment by daniel-luberda** [comment]
@chrisfoulds I'll need more info. 2.0.9 is quite old, is possible to have some sample project which demonstrates what is not working for you?

**Comment by daniel-luberda** [comment]
@chrisfoulds Also try to disable Android fast renderer (it's enabled by default if supported). 

`CachedImageRenderer.Init(false)`

**Comment by EmilAlipiev** [comment]
hi @daniel-luberda, i am having problem with 2.2.3 tap gesture doesnt work in some cases. I am using syncfusion listview and it has swipe action and within swipe I have ffimageloading image. it used to work fine but latest version doesnt work. I replaced with XF image. it works fine. Do you have any clue? otherwise, I will try to get you a small repro project. do you prefer another issue for this?

**Comment by daniel-luberda** [comment]
Fixed in 2.3 - it's now a default behaviour, so `CachedImage` behaviour is the same as `Image`

@EmilAlipiev Did you try to disable fast renderer? Please create a separate issue.

**Comment by EmilAlipiev** [comment]
@daniel-luberda  i am not experiencing this issue anymore. I dont know what fixed but using latest XF and FFimageloading. i dont have it today. thank you

**Comment by swansca79** [comment]
Sorry to beat a dead horse, but this is not fixed on iOS.  I just updated to the latest FFImageLoading version today, from 2.2.3, also on the latest Xamarin Forms and iOS as of right now.

Black is Image, Blue is CachedImage.
<img width="380" alt="screen shot 2018-01-04 at 5 12 39 pm" src="https://user-images.githubusercontent.com/17572520/34588418-8c912ed4-f172-11e7-950d-9c0b04a204e8.png">


`
2018-01-04 17:02:53.905 VideoChat.iOS[341:46667] DownsampleToViewSize failed - view is expandable in both dimensions, so it doesn't have a size. Please use DownsampleWidth or DownsampleHeight property.
`

I get this message a lot now, is it expected that a CachedImage must have DownsideWidth/Height set all the time?  Even when I set them, the icons are still taking up all available space.

I'll put together a simple sample of this that I can share soon

**Comment by daniel-luberda** [comment]
@swansca79 

- Did you manually set `CachedImage.FixedOnMeasureBehavior`?
- What is your image code? 

> I get this message a lot now, is it expected that a CachedImage must have DownsideWidth/Height set all the time? Even when I set them, the icons are still taking up all available space.

Error is self explanatory, read more here: https://github.com/luberda-molinet/FFImageLoading/wiki/Xamarin.Forms-API#downsampletoviewsize-bool-default-false

Because of this, all your images are downsampled to a fallback value = `100px`. That is the reason all images have the same size.

**Comment by swansca79** [comment]
I did not specifically change the FixedOnMeasureBehavior, which should be true according to the docs?  The Downssample-width/height properties don't appear to do anything without WidthRequest or HeightRequest.  If I set WidthRequest=24, then it works, but that is redundant since it's the same size of the image.  I don't feel like having to manually set size properties on every image is a solution especially since the standard Image tag will respect the original size.

**Comment by daniel-luberda** [comment]
> I did not specifically change the FixedOnMeasureBehavior, which should be true according to the docs? 

Yes, it's enabled by default now.

> The Downssample-width/height properties don't appear to do anything without WidthRequest or HeightRequest. If I set WidthRequest=24, then it works, but that is redundant since it's the same size of the image.

@swansca79 Could you make a repo?

**Comment by swansca79** [comment]
I'll set one up

**Comment by swansca79** [comment]
Repo: https://github.com/swansca79/FFImageLoadingInvestigation

I added a few combinations of the width/downsample attributes

**Comment by daniel-luberda** [comment]
@swansca79 Thank you for a repo, it was very useful for me.

**Comment by swansca79** [comment]
Great, thanks!

**Comment by maxyhi** [comment]
this doesn't seem to work for Svg Cached Image
